File: event-sourcing/csharp/src/EventSourcing/Events.cs
namespace EventSourcing;

public abstract record AccountEvent(string AccountId, DateTime Timestamp);

public record AccountOpened(string AccountId, string OwnerName, DateTime Timestamp)
    : AccountEvent(AccountId, Timestamp);

public record MoneyDeposited(string AccountId, Money Amount, DateTime Timestamp)
    : AccountEvent(AccountId, Timestamp);

public record MoneyWithdrawn(string AccountId, Money Amount, DateTime Timestamp)
    : AccountEvent(AccountId, Timestamp);

public record AccountClosed(string AccountId, DateTime Timestamp)
    : AccountEvent(AccountId, Timestamp);

File: event-sourcing/python/src/event_sourcing/events.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime

from .money import Money


@dataclass(frozen=True)
class AccountEvent:
    account_id: str
    timestamp: datetime


@dataclass(frozen=True)
class AccountOpened(AccountEvent):
    owner_name: str = ""


@dataclass(frozen=True)
class MoneyDeposited(AccountEvent):
    amount: Money = Money.zero()


@dataclass(frozen=True)
class MoneyWithdrawn(AccountEvent):
    amount: Money = Money.zero()


@dataclass(frozen=True)
class AccountClosed(AccountEvent):
    pass
